pages/workspaceLogs/workspaceLogs.js

```diff
@@ -0,0 +1,168 @@
+// @ts-check
+const express = require('express');
+const router = express.Router();
+const asyncHandler = require('express-async-handler');
+const AWS = require('aws-sdk');
+const z = require('zod');
+
+const config = require('../../lib/config');
+const error = require('../../prairielib/lib/error');
+const sqldb = require('../../prairielib/lib/sql-db');
+const sqlLoader = require('../../prairielib/lib/sql-loader');
+
+const { WorkspaceLogs, WorkspaceVersionLogs } = require('./workspaceLogs.html');
+
+const sql = sqlLoader.loadSqlEquiv(__filename);
+
+/**
+ * Loads logs for the given workspace version from S3. If logs cannot be found,
+ * returns null.
+ *
+ * @param {string | number} workspaceId
+ * @param {string | number} version
+ * @param {string} [startAfter]
+ *
+ * @returns {Promise<{ logs: string, startAfter: string} | null>}
+ */
+async function loadLogParts(workspaceId, version, startAfter) {
+  // TODO: handle errors and missing data.
```

Comment:
I think this is done - have I missed anything?
```suggestion
```

Reply:
Hopefully all this complexity will go away if you shift to the "pull from workspace_host, fall back to S3" scheme.

Reply:
Yes, it should!

Reply:
And indeed it did 🎉

Reply:
So nice!